guides/cookbook.md

```diff
@@ -391,6 +391,217 @@ The result of `getBytes()` from a text or binary writer can subsequently
+
+To write binary Ion instead of text, use the `binary=True` option.
+
+```python
+ion = BytesIO()
+simpleion.dump(value, ion, binary=True)
+print(ion.getvalue())
+```
+
+Result: `b'\xe0\x01\x00\xea\xec\x81\x83\xde\x88\x87\xb6\x85hello\xde\x87\x8a\x85world'`
+
+When reading Ion streams that contain multiple top-level values, provide the `single_value=False`
+option to `load/loads` to receive all of the values within a Python `list`.
+
+```python
+data = u'1 2 3'
+value = simpleion.loads(data, single_value=False)
+print(value)
+```
+
+Result: `[1, 2, 3]`
+
+To write a sequence type as a stream of top-level values, provide the `sequence_as_stream=True`
+option to `dump/dumps`.
+
+```python
+print(simpleion.dumps(value, sequence_as_stream=True, binary=False))
+```
+
+Result: `$ion_1_0 1 2 3`
+
+The `simpleion` [API documentation](https://ion-python.readthedocs.io/en/latest/amazon.ion.html#module-amazon.ion.simpleion)
+enumerates the complete set of options available.
+
+#### Using events
+
+The non-blocking event-based APIs are useful for streaming reading and writing of Ion data. This
+enables use cases where data becomes available incrementally or only needs to be sparsely parsed.
+
+```python
+from amazon.ion.reader import NEXT_EVENT, read_data_event
+from amazon.ion.reader_managed import managed_reader
+from amazon.ion.reader_text import text_reader
+
+# Create a text reader coroutine that manages symbol tables
+# automatically.
+reader = managed_reader(text_reader())
+event = reader.send(NEXT_EVENT)
+# No data has been provided, so the reader is at STREAM_END
+# and will wait for data.
+assert event.event_type == IonEventType.STREAM_END
+# Send an incomplete Ion value.
+event = reader.send(read_data_event(b'{hello:'))
+# Enough data was available for the reader to determine that
+# the start of a struct value has been encountered.
+assert event.event_type == IonEventType.CONTAINER_START
+assert event.ion_type == IonType.STRUCT
+# Advancing the reader causes it to step into the struct.
+event = reader.send(NEXT_EVENT)
+# The reader reached the end of the data before completing
+# a value. Therefore, an INCOMPLETE event is returned.
+assert event.event_type == IonEventType.INCOMPLETE
+# Send the rest of the value.
+event = reader.send(read_data_event(b'\"world\"}'))
+# The reader now finishes parsing the value within the struct.
+assert event.event_type == IonEventType.SCALAR
+assert event.ion_type == IonType.STRING
+hello = event.field_name.text
+world = event.value
+# Advance the reader past the string value.
+event = reader.send(NEXT_EVENT)
+# The reader has reached the end of the struct.
+assert event.event_type == IonEventType.CONTAINER_END
+# Advancing the reader causes it to step out of the struct.
+event = reader.send(NEXT_EVENT)
+# There is no more data and a value has been completed.
+# Therefore, the reader conveys STREAM_END.
+assert event.event_type == IonEventType.STREAM_END
+print(u'%s %s' % (hello, world))
```

Comment:
Good walk-through. I would add an empty line between steps to make them visually easier to consume.

Reply:
I tried this, but thought it was too much. Fortunately, the rendered view creates visual separation because the comments are a light shade of grey while the code is syntax highlighted.